docs/type_promotion.rst

```diff
@@ -133,6 +133,13 @@ in the table above. There are three key differences:
 * when promoting a Python scalar value against a typed JAX value of the same category,
   JAX always prefers the precision of the JAX value. For example, ``jnp.int16(1) + 1``
   will return ``int16`` rather than promoting to ``int64`` as in Numpy.
+  Note that this applies only to Python scalar values; if the constant is a NumPy
+  array then the above lattice is used for type promotion.
+  For example, ``jnp.int16(1) + np.array(1)``
+  will return ``int64``. This may result in confusing non-associative behavior:
+  ``np.int16(1) +jnp.int16(2) + 3`` results in a ``int16``, while
+  ``3 + np.int16(1) + jnp.int16(2)`` results in an ``int64``, because the result
+  of ``3 + np.int16(1)`` is a NumPy array of dtype ``int64``.
 
 * when promoting an integer or boolean type against a floating-point or complex
```

Comment:
Nit suggestion for readability:

```suggestion
* When promoting a Python scalar value against a typed JAX value of the same category,
  JAX always prefers the precision of the JAX value. For example, ``jnp.int16(1) + 1``
  will return ``int16`` rather than promoting to ``int64`` as
  in NumPy. Note that this applies only to Python scalar values; if the constant is a NumPy
  array then the above lattice is used for type promotion.
  For example, ``jnp.int16(1) + np.array(1)``
  will return ``int64``. This may result in confusing non-associative behavior:
  ``np.int16(1) +jnp.int16(2) + 3`` results in a ``int16``, while
  ``3 + np.int16(1) + jnp.int16(2)`` results in an ``int64``, because the result
  of ``3 + np.int16(1)`` is a NumPy array of dtype ``int64``.

* When promoting an integer or boolean type against a floating-point or complex
```

Maybe there's a way to highlight that _JAX always prefers the precision of the JAX value_  and/or _this applies only to Python scalar values_ (**bolden then?**) since, for example, 3 + int16s is an int64 (!) and that's a JAX Gotcha 🔪 that may catch some NumPy users off-guard.

Reply:
@8bitmp3 Since I have more changes to make on my local copy, I have just incorporated your suggestions into my changes, rather than committing them (I do not know of an easy process by which to merge the committed suggestions and my subsequent local changes).